[test_ring_joint_sdpa.py] | Blackhole sanity tests | blackhole-multi-card-fast-sanity-tests / ring sdpa PERF tests (QB2 only) [bh_quietbox_2]

## CI Failure Report

<!-- AUTO-TRIAGE-SUMMARY-START -->
**Summary:** [CI] ring SDPA perf tests fail: "Real-time profiler must be active for SDPA perf checks"
<!-- AUTO-TRIAGE-SUMMARY-END -->

**Workflow:** Blackhole sanity tests
**Job:** blackhole-multi-card-fast-sanity-tests / ring sdpa PERF tests (QB2 only) [bh_quietbox_2]

### Regression timeline

- **Last passing:** commit `019491d78ecc` — 2026-07-10T16:46:07Z — [run](https://github.com/tenstorrent/tt-metal/actions/runs/29108573197)
- **First failing:** commit `1af2e531999e` — 2026-07-11T00:47:52Z — [job](https://github.com/tenstorrent/tt-metal/actions/runs/29133371422/job/86496525296)
- **Most recent failing:** commit `df6228ba8a48` — 2026-07-11T12:26:30Z — [job](https://github.com/tenstorrent/tt-metal/actions/runs/29152630758/job/86546205965)

### Failing job URLs (last 4 runs)

- https://github.com/tenstorrent/tt-metal/actions/runs/29152630758/job/86546205965
- https://github.com/tenstorrent/tt-metal/actions/runs/29146484003/job/86529992334
- https://github.com/tenstorrent/tt-metal/actions/runs/29140739768/job/86515267393
- https://github.com/tenstorrent/tt-metal/actions/runs/29133371422/job/86496525296

### Failing test path(s)

- `tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_perf_check[wan2_2_1xGLX-q288-k512-ring4]`
- `tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_perf_check[mla_100k-q160-k320-ring4]`

### Error excerpt (terminal failure from logs)

```
FAILED tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_perf_check[wan2_2_1xGLX-q288-k512-ring4] - Failed: Real-time profiler must be active for SDPA perf checks
FAILED tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_perf_check[mla_100k-q160-k320-ring4] - Failed: Real-time profiler must be active for SDPA perf checks
```

The pytest assertion raised in both cases:
```python
pytest.fail("Real-time profiler must be active for SDPA perf checks")
```

### Reproduction steps

Run the failing tests on a Blackhole multi-card setup (bh_quietbox_2) with the real-time profiler **disabled**:

```bash
pytest tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py \
  -k "test_ring_joint_attention_perf_check" \
  --timeout=300
```

To reproduce the pass condition, ensure the real-time profiler is active (e.g., `TT_METAL_DEVICE_PROFILER=1` or equivalent fixture/env var) before running.

### Notes

- The test explicitly calls `pytest.fail("Real-time profiler must be active for SDPA perf checks")` when no real-time profiler is detected, indicating the profiler guard was either newly added or the CI environment stopped enabling it between the last passing commit (`019491d78ecc`) and the first failing commit (`1af2e531999e`).
- Run 3 and Run 4 logs include an advisory message: *"Enable the real-time profiler before running SDPA perf check tests — ensure `TT_METAL_DEVICE_PROFILER=1` or equivalent profiler activation flag is set in the CI environment for this test job."*
- Exactly 2 tests fail in every run, both under the same `test_ring_joint_attention_perf_check` parametrized test.

**Determinism check:**
- Run 1 (df6228ba8a48): `FAILED tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_perf_check[wan2_2_1xGLX-q288-k512-ring4] - Failed: Real-time profiler must be active for SDPA perf checks`
- Run 2 (commit ~07-11T08:59): `FAILED tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_perf_check[wan2_2_1xGLX-q288-k512-ring4] - Failed: Real-time profiler must be active for SDPA perf checks`
- Run 3 (commit ~07-11T06:01): `FAILED tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_perf_check[wan2_2_1xGLX-q288-k512-ring4] - Failed: Real-time profiler must be active for SDPA perf checks`
- Run 4 (1af2e531999e): `FAILED tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_perf_check[wan2_2_1xGLX-q288-k512-ring4] - Failed: Real-time profiler must be active for SDPA perf checks`

<!-- AUTO-TRIAGE-METADATA-START -->
`Auto-triage-workflow: Blackhole sanity tests`
`Auto-triage-job-name: blackhole-multi-card-fast-sanity-tests / ring sdpa PERF tests (QB2 only) [bh_quietbox_2]`
<!-- AUTO-TRIAGE-METADATA-END -->

<!-- AUTO-TRIAGE-INFO-START -->
## About this issue (automated CI auto-triage)
This issue was opened and assigned by an automated AI system. It posts through a bot token, so notifications may look like they come from a person, but they are automated.

The owner is selected from recent commit history and job-ownership data. It reflects who is *likely able to fix or route* this, not necessarily who caused it, so it can be imperfect.

**If this was assigned to the wrong person, you can:**
- Comment `REASSIGN OWNER: <why this is outside your area / which team or test-area it belongs to>` and the system will automatically re-route it to a better-suited owner.
- Unassign yourself (this stops notifications to you).
- Close the issue if it is not a real or relevant failure.

The assigned owner is reminded about every 3 days until the issue is resolved. Full guide: https://tenstorrent.atlassian.net/wiki/spaces/MI6/pages/2545221692/Auto-Triage+Info
<!-- AUTO-TRIAGE-INFO-END -->

<!-- PIPELINE-REPORT-START -->
## Pipeline report
Analysis report: https://github.com/tenstorrent/tt-metal/actions/runs/28900475665
Slack notification: https://tenstorrent.slack.com/archives/C0B1F8Z7GQM/p1783461040389919?thread_ts=1783460260.483619&cid=C0B1F8Z7GQM
<!-- PIPELINE-REPORT-END -->

--- a/.github/time_budget.yaml
+++ b/.github/time_budget.yaml
@@ -87,7 +87,7 @@ ttnn:
     bh_p100: 545
     bh_p150b_civ2: 545
     bh_quietbox_2: 35
-    bh_quietbox_2_iommu: 20 # sdpa nightly tests moved here from bh_quietbox_2 (IOMMU-enabled QB2)
+    bh_quietbox_2_iommu: 30 # sdpa nightly (20) + ring sdpa PERF (10); both need IOMMU for the RT profiler (IOMMU-enabled QB2)
     bh_p300: 15
     wh_llmbox: 25
     sim_wormhole_b0: 680

--- a/tests/nightly/blackhole/sdpa/test_exp_ring_joint_sdpa.py
+++ b/tests/nightly/blackhole/sdpa/test_exp_ring_joint_sdpa.py
@@ -22,7 +22,7 @@ from loguru import logger
 from ttnn.operations.ccl import Topology
 
 import ttnn
-from models.common.utility_functions import skip_with_llk_assert
+from models.common.utility_functions import skip_with_llk_assert, skip_with_watcher
 from tests.nightly.sdpa_perf_utils import (
     compute_cores_used,
     compute_math_utilization,
@@ -821,6 +821,7 @@ EXP_RING_JOINT_PERF_CHECK_CONFIGS = [
     ids=[f"ring{cfg[0]}-{cfg[2]}" for cfg in EXP_RING_JOINT_PERF_CHECK_CONFIGS],
 )
 @skip_with_llk_assert("No need to verify LLK asserts for performance tests.")
+@skip_with_watcher("Watcher perturbs kernel timing; perf checks are not meaningful with it enabled.")
 def test_exp_ring_joint_attention_perf_check(ring_size_expected, max_payload_size, payload_id, expected_util):
     """Measure exp ring joint SDPA math utilization via real-time device program records."""
     num_devices = detect_devices_without_opening()

--- a/tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py
+++ b/tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py
@@ -33,7 +33,7 @@ from loguru import logger
 from ttnn.operations.ccl import Topology
 
 import ttnn
-from models.common.utility_functions import skip_with_llk_assert
+from models.common.utility_functions import skip_with_llk_assert, skip_with_watcher
 
 # ============================================================================
 # CONFIGURATION CONSTANTS
@@ -3503,6 +3503,7 @@ else:
     ids=[f"{cfg[0]}-q{cfg[1]}-k{cfg[2]}-ring{cfg[3]}" for cfg in RING_JOINT_PERF_CHECK_CONFIGS],
 )
 @skip_with_llk_assert("No need to verify LLK asserts for performance tests.")
+@skip_with_watcher("Watcher perturbs kernel timing; perf checks are not meaningful with it enabled.")
 def test_ring_joint_attention_perf_check(
     model_name, q_chunk_size, k_chunk_size, ring_size_expected, expected_util, margin
 ):
@@ -3557,6 +3558,8 @@ def test_ring_joint_attention_perf_check(
     )
 
 
+@skip_with_llk_assert("No need to verify LLK asserts for performance tests.")
+@skip_with_watcher("Watcher perturbs kernel timing; perf checks are not meaningful with it enabled.")
 def test_ring_mla_perf_better_than_separate_v_ring_joint():
     """Profile ring_mla against the existing separate-V MLA ring joint path and require a speedup."""
     model_name = "mla_100k"
@@ -4361,6 +4364,7 @@ else:
     ids=[f"{cfg[0]}-q{cfg[1]}-k{cfg[2]}-ring{cfg[3]}" for cfg in RING_MLA_CHUNKED_PERF_CHECK_CONFIGS],
 )
 @skip_with_llk_assert("No need to verify LLK asserts for performance tests.")
+@skip_with_watcher("Watcher perturbs kernel timing; perf checks are not meaningful with it enabled.")
 def test_ring_mla_chunked_perf_check(model_name, q_chunk_size, k_chunk_size, ring_size_expected, expected_util):
     """Measure ring_mla chunked-prefill math utilization for the kimi 50k+5k galaxy chunk (a 5k Q
     chunk against a 50k K/V prefix), simulated on the 4-device QuietBox, via realtime profiler and assert
@@ -4424,6 +4428,8 @@ def test_ring_mla_chunked_perf_check(model_name, q_chunk_size, k_chunk_size, rin
     MINIMAX3_GQA_CHUNKED_PERF_CHECK_CONFIGS,
     ids=[f"{cfg[0]}-q{cfg[1]}-k{cfg[2]}-ring{cfg[3]}" for cfg in MINIMAX3_GQA_CHUNKED_PERF_CHECK_CONFIGS],
 )
+@skip_with_llk_assert("No need to verify LLK asserts for performance tests.")
+@skip_with_watcher("Watcher perturbs kernel timing; perf checks are not meaningful with it enabled.")
 def test_ring_joint_attention_minimax3_gqa_chunked_perf_check(
     model_name, q_chunk_size, k_chunk_size, ring_size_expected, expected_util
 ):

--- a/tests/nightly/blackhole/sdpa/test_scaled_dot_product_attention_sprint.py
+++ b/tests/nightly/blackhole/sdpa/test_scaled_dot_product_attention_sprint.py
@@ -13,6 +13,7 @@ from tests.tt_eager.python_api_testing.sweep_tests.comparison_funcs import (
 import ttnn
 from loguru import logger
 import pytest
+from models.common.utility_functions import skip_with_llk_assert, skip_with_watcher
 
 
 def fa_rand(*shape):
@@ -435,6 +436,8 @@ SDPA_PERF_CHECK_CONFIGS = [
     SDPA_PERF_CHECK_CONFIGS,
     ids=[f"{cfg[0]}-q{cfg[1]}-k{cfg[2]}" for cfg in SDPA_PERF_CHECK_CONFIGS],
 )
+@skip_with_llk_assert("No need to verify LLK asserts for performance tests.")
+@skip_with_watcher("Watcher perturbs kernel timing; perf checks are not meaningful with it enabled.")
 def test_sdpa_perf_check(sdpa_realtime_profiled_device, shape_id, q_chunk_size, k_chunk_size, expected_util):
     """Measure single-chip SDPA math utilization via real-time device program records."""
     idx = INPUT_IDS.index(shape_id)

--- a/tests/pipeline_reorg/blackhole_multi_card_sanity_tests.yaml
+++ b/tests/pipeline_reorg/blackhole_multi_card_sanity_tests.yaml
@@ -55,13 +55,18 @@
   owner_id: U05ACKAJTHS # Naif Tarafdar
   team: ttnn
 
-- name: ring sdpa PERF tests (QB2 only)
+- name: ring sdpa PERF tests (QB2 IOMMU only)
+  # These perf checks measure device kernel duration via the realtime device profiler. On Blackhole
+  # the RT profiler's D2H socket uses 64-bit PCIe addressing, which requires host IOMMU (there is no
+  # hugepage fallback) — see evaluate_realtime_profiler_eligibility in realtime_profiler_manager.cpp.
+  # Without IOMMU the profiler never activates and the tests fail the guard (issue #49665), so this
+  # job must run on the IOMMU-enabled runner, matching the sdpa nightly job above.
   cmd: |
     pytest tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_joint_attention_perf_check -svv
     pytest tests/nightly/blackhole/sdpa/test_ring_joint_sdpa.py::test_ring_mla_chunked_perf_check -svv
     pytest tests/nightly/blackhole/sdpa/test_exp_ring_joint_sdpa.py::test_exp_ring_joint_attention_perf_check -svv
   skus:
-    bh_quietbox_2:
+    bh_quietbox_2_iommu:
       timeout: 10
   owner_id: U05ACKAJTHS # Naif Tarafdar
   team: ttnn

